### PR TITLE
Feat/Intelligent highlight

### PR DESCRIPTION
**Pull request by solid-danylokhvan**
https://github.com/solid-software/languagetool_textfield/assets/113029066/79723e86-c629-4a8e-a79a-27e98f67e2c2

### CODE REVIEW

**Review by solid-yuriiprykhodko — CHANGES_REQUESTED** [review]
Please add a test plan, and provide testing results.

Off the top of my head, we should test:
- copy-pasting words/sentences
- editing test in the beginning, middle and end of the text
- selecting and removing large chunks of text
- combinations of the above

For all these cases, we must ensure that the custom styling and interactivity works appropriately.

**Comment by solid-danylokhvan** [comment]
**TEST REPORT**

1. copy-pasting words/sentences

https://github.com/solid-software/languagetool_textfield/assets/113029066/e4b9bf02-133a-4050-abc2-5ea0b17969b2

2. editing text in the beginning, middle and end of the text

https://github.com/solid-software/languagetool_textfield/assets/113029066/0c4f18ce-0263-461e-8641-0c03d2606b2e

3. selecting and removing large chunks of text

https://github.com/solid-software/languagetool_textfield/assets/113029066/967cc48a-8d6b-4373-9e45-b4c6e58bf5f3

4. combinations of the above

https://github.com/solid-software/languagetool_textfield/assets/113029066/af06d6e1-2516-423e-970e-200175ec05f9

File: lib/core/controllers/colored_text_editing_controller.dart [hunk 1]
```diff
@@ -68,33 +71,51 @@ class ColoredTextEditingController extends TextEditingController {
...
     _recognizers.clear();
 
-    final mistakesWrapper = await languageCheckService.findMistakes(newText);
+    // Set the flag to indicate an ongoing request
+    // Increment the request ID to track the most recent request
+    final requestId = ++_lastRequestId;
+
+    // Find mistakes in the new text using the language check service
+    languageCheckService.findMistakes(newText).then((value) {
+      // Retrieve the result and error information from the MistakesWrapper
```
Comment (solid-yuriiprykhodko) [inline lib/core/controllers/colored_text_editing_controller.dart:104]: Redundant comments.
```suggestion
    languageCheckService.findMistakes(newText).then((value) {
```

Comment (solid-yuriiprykhodko) [reply lib/core/controllers/colored_text_editing_controller.dart:104]: Also, no good reason to use Future.then callbacks here. Let's revert to `async-await` and reduce nesting.

File: lib/core/controllers/colored_text_editing_controller.dart [hunk 2]
```diff
@@ -68,33 +71,51 @@ class ColoredTextEditingController extends TextEditingController {
...
+      final mistakesWrapper = value;
+      final mistakes = mistakesWrapper?.result();
+      _fetchError = mistakesWrapper?.error;
 
-    _mistakes =
-        mistakesWrapper.hasResult ? mistakesWrapper.result().toList() : [];
-    _fetchError = mistakesWrapper.error;
+      // Check if a newer request has been initiated during the API call
+      if (requestId != _lastRequestId) return;
+      _lastRequestId = 0;
```
Comment (solid-yuriiprykhodko) [inline lib/core/controllers/colored_text_editing_controller.dart:111]: This shouldn't be done inside this class.

We want to update the UI using the latest request result, correct? Sounds like we need another `LanguageToolService` modifier for this, similar to how we implement debounce -- one that would discard stale results and return null.

File: lib/core/controllers/colored_text_editing_controller.dart [hunk 3]
```diff
@@ -158,6 +179,90 @@ class ColoredTextEditingController extends TextEditingController {
     );
   }
 
+  /// Filters the list of mistakes based on the changes
+  /// in the text when it is changed.
+  List<Mistake> _filterMistakesOnChanged(String newText) {
```
Comment (solid-yuriiprykhodko) [inline lib/core/controllers/colored_text_editing_controller.dart:184]: This function needs some decomposition.
A good rule of thumb is no more than 1 level of code nesting per function (e.g. a single loop, or a single if/if-else condition). 
Here we have 3, with  `for - if - if`.

Selection handling is unreadable -- maybe we need a set of utility functions or a class that would encapsulate these comparisons.

**Review by solid-yuriiprykhodko — CHANGES_REQUESTED** [review]

File: lib/core/controllers/colored_text_editing_controller.dart (hunk 2 above)
Comment (solid-danylokhvan) [reply lib/core/controllers/colored_text_editing_controller.dart]: I've added the request id field to the debounce class

File: lib/core/controllers/colored_text_editing_controller.dart [hunk 4]
```diff
@@ -190,6 +207,45 @@ class ColoredTextEditingController extends TextEditingController {
...
+      final isTextLengthNotIncreasedAndBaseBeforeOrAtMistake =
+          !isLengthIncreased && mistakeRange.isBeforeOrAt(baseOffset);
+
+      final lengthDiscrepancy = newText.length - text.length;
+      final newOffset = mistake.offset + lengthDiscrepancy;
+
+      isTextLengthIncreasedAndBaseBeforeMistake ||
+              isTextLengthNotIncreasedAndBaseBeforeOrAtMistake
+          ? newMistakes.add(mistake.copyWith(offset: newOffset))
+          : newMistakes.add(mistake);
```
Comment (solid-yuriiprykhodko) [inline lib/core/controllers/colored_text_editing_controller.dart:243]: 1. Let's turn this into a generator function.
2. selection.base.offset is going to be the location of the _caret_, correct? Having no selection range is probably going to be our most common use case, perhaps handling just a plain _caret cursor_ and a selection _range_ should be handled differently?

File: lib/implementations/keep_latest_response_service.dart [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+
+  /// Executes the latest operation and returns its result.
+  /// Only the results of the most recent operation are returned,
+  /// discarding any previous ongoing operations.
+  Future<T?> processLatestOperation<T>(Future<T> Function() action) async {
+    await _currentOperation?.cancel();
+    _currentOperation = CancelableOperation<T>.fromFuture(action());
+
+    return _currentOperation?.value as Future<T>?;
+  }
```
Comment (solid-yuriiprykhodko) [inline lib/implementations/keep_latest_response_service.dart:16]: To provide best service, we need to kick off the `action` _before_ we do anything else.
Let's reorder things:
1. Run action
2. cancel previous operation if needed
3. store pointer to current action

**Review by solid-yuriiprykhodko — CHANGES_REQUESTED** [review]

File: lib/core/controllers/colored_text_editing_controller.dart [hunk 6]
```diff
@@ -190,6 +206,66 @@ class ColoredTextEditingController extends TextEditingController {
...
+  /// Adjusts the mistake offset when the selection is a caret cursor.
+  Mistake? _adjustMistakeOffsetWithCaretCursor({
+    required Mistake mistake,
+    required int lengthDiscrepancy,
+  }) {
+    final mistakeRange = ClosedRange(mistake.offset, mistake.endOffset);
+    final baseOffset = selection.base.offset;
+    final shouldSkipOffsetAdjustment = mistakeRange.contains(baseOffset);
+
+    if (shouldSkipOffsetAdjustment) return null;
```
Comment (solid-yuriiprykhodko) [inline lib/core/controllers/colored_text_editing_controller.dart:241]:
```suggestion
    final caretLocation = selection.base.offset;
    final isCaretOnMistake = mistakeRange.contains(caretLocation);
    
    // Don't highlight mistakes on changed text until we get an update from the API.
    if (isCaretOnMistake) return null;
```

File: lib/core/controllers/colored_text_editing_controller.dart [hunk 7]
```diff
@@ -190,6 +206,66 @@ class ColoredTextEditingController extends TextEditingController {
...
+    required int lengthDiscrepancy,
+  }) {
+    final selectionRange = ClosedRange(selection.start, selection.end);
+    final mistakeRange = ClosedRange(mistake.offset, mistake.endOffset);
+
+    final shouldSkipOffsetAdjustment =
+        selectionRange.containsRange(mistakeRange) ||
+            selectionRange.overlapsWith(mistakeRange);
+
+    if (shouldSkipOffsetAdjustment) return null;
```
Comment (solid-yuriiprykhodko) [inline lib/core/controllers/colored_text_editing_controller.dart:261]: 1. `ClosedRange::overlapsWith` is always true if `ClosedRange::containsRange` is true.
2. Descriptive naming
```suggestion
    final hasSelectedTextChanged = selectionRange.overlapsWith(mistakeRange);

    if (hasSelectedTextChanged) return null;
```

File: lib/implementations/keep_latest_response_service.dart [hunk 8]
```diff
@@ -0,0 +1,18 @@
+import 'package:async/async.dart';
```
Comment (solid-yuriiprykhodko) [inline lib/implementations/keep_latest_response_service.dart:1]: Move this to /utils. This is not an implementation of a LangToolService.

**Review by solid-yuriiprykhodko — CHANGES_REQUESTED** [review]
A lot better and easy to reason about.
A few more readability changes and we're good to go.

**Review by solid-yuriiprykhodko — APPROVED** [review]
LGTM!

**Pull request merged**